File: backend/app/models/maintenance.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column

from ..db.base import Base


class MaintenancePlan(Base):
    __tablename__ = "maintenance_plan"

    plan_id: Mapped[str] = mapped_column(String(32), primary_key=True)
    device_id: Mapped[str] = mapped_column(ForeignKey("device_info.device_id", ondelete="CASCADE", onupdate="CASCADE"), index=True)
    plan_start_time: Mapped[datetime] = mapped_column(DateTime)
    plan_end_time: Mapped[datetime] = mapped_column(DateTime)
    maintenance_type: Mapped[str | None] = mapped_column(String(16))
    duration_minutes: Mapped[int | None] = mapped_column(Integer)
    deadline: Mapped[datetime | None] = mapped_column(DateTime)
    risk_level: Mapped[str | None] = mapped_column(String(16))
    source: Mapped[str] = mapped_column(String(16), default="auto")
    reason: Mapped[str | None] = mapped_column(String(128))
    status: Mapped[str] = mapped_column(String(16), default="pending", index=True)
    create_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    update_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

File: frontend/src/types/maintenance.ts
export interface MaintenancePlan {
  plan_id: string;
  device_id: string;
  plan_start_time: string;
  plan_end_time: string;
  maintenance_type: string | null;
  duration_minutes: number | null;
  deadline: string | null;
  risk_level: string | null;
  source: string;
  reason: string | null;
  status: string;
  create_time: string | null;
  update_time: string | null;
}

export interface MaintenanceWindowGenerateRequest {
  device_ids: string[];
}

export interface MaintenancePlanUpdateRequest {
  plan_start_time?: string | null;
  plan_end_time?: string | null;
  maintenance_type?: "repair" | "replace" | null;
  risk_level?: "low" | "medium" | "high" | null;
  reason?: string | null;
  status?: "pending" | "confirmed" | "executing" | "finished" | "cancelled" | null;
}

export interface MaintenanceGenerateItem {
  device_id: string;
  generated: boolean;
  message: string;
  plan: MaintenancePlan | null;
  conflict_task_ids: string[];
}

export interface MaintenanceGenerateResponse {
  items: MaintenanceGenerateItem[];
}

export interface MaintenanceGanttItem {
  id: string;
  device_id: string;
  title: string;
  start_time: string;
  end_time: string;
  item_type: string;
  status: string;
  conflict_task_ids: string[];
}

export interface MaintenanceGantt {
  items: MaintenanceGanttItem[];
}
